frontend/src/components/data-table/column-pinning/feature.ts

```diff
@@ -0,0 +1,194 @@
+          newState = {
+            ...old,
+            [prevValue]: old[prevValue]?.filter((id) => id !== column.id),
+            [position]: [...(old[position] || []), column.id],
+          };
+        }
+
+        // If only "select" column is pinned to the left, unpin it
+        if (
+          position !== "left" &&
+          newState.left?.length === 1 &&
+          newState.left[0] === "select"
+        ) {
+          // unpin the 'select' column
+          table.getColumn("select")?.pin(false);
+
+          // remove the 'select' column from the left side
+          newState = {
+            ...newState,
+            left: [],
+          };
+        }
+
+        return newState;
+      };
+      table.options.onColumnPinningChange?.(safeUpdater);
+    };
+
+    // toggle the column's pinning state
+    column.toggleColumnPinning = (position: ColumnPinningPosition) => {
+      // get the column pinning state
+      const { columnPinning } = table.getState();
+
+      // if the column is being pinned to the left,
+      // but the 'select' column is not pinned to the left,
+      // pin the 'select' column to the left
+      if (position === "left" && !columnPinning.left?.includes("select")) {
+        // pin the 'select' column to the left
+        const selectColumn = table.getColumn("select");
+        if (selectColumn) {
+          // always setColumnPinning first then pin
+          selectColumn.setColumnPinning("left");
+          selectColumn.pin("left");
+        }
+      }
+      column.setColumnPinning(position);
+      column.pin(position);
+    };
+
+    // get the column's common pinning styles
+    column.getCommonPinningStyles = (): CSSProperties => {
+      // get the column's pinning state
+      const isPinned = column.getIsPinned();
+
+      // get if the column is the last pinned column on the left
+      const isLastLeftPinnedColumn =
+        isPinned === "left" &&
+        column.getIsLastColumn("left") &&
+        // don't apply the box shadow to the 'select' column
+        column.id !== "select";
+
+      // get if the column is the first pinned column on the right
+      const isFirstRightPinnedColumn =
+        isPinned === "right" && column.getIsFirstColumn("right");
+
+      return {
+        // apply a box shadow to the column if
+        // it is the last pinned column on the left
+        // or the first pinned column on the right
+        boxShadow: isLastLeftPinnedColumn
+          ? "-4px 0 4px -4px gray inset"
+          : isFirstRightPinnedColumn
+            ? "4px 0 4px -4px gray inset"
+            : undefined,
+
+        // set the column's width
+        width: column.getSize(),
+
+        // set the column's left position
+        left:
+          isPinned === "left"
+            ? `${
+                column.getStart("left") -
+                (column.id === "select" ? 0 : (column.getSize() * 4) / 5)
+              }px`
+            : undefined,
```

Comment:
there seems to be a small gap between columns with this method:

![image](https://github.com/user-attachments/assets/8ee9f4f9-27eb-4014-ae6b-e95dfd8c233a)

Reply:
you could use the column ordering API in here instead. 

https://tanstack.com/table/v8/docs/api/features/column-ordering

still keep the same left/right state, but after each pin/unpin, you can update the ordering:

```
const allColumnIds = table.getColumnIds()
const unsetColumnIds = allColumnIds.filter(id => `!left.includes(id) && !right.includes(id)`)
return table.updateColumnOrder([..left, ...unsetColumnIds, ...right]);
```

then you wont need to mess with `width` and `sizing` and you can just use `sticky`

Reply:
actually we can disregard this last message, i think what you have is close enough if we can re-use the existing table feature instead of building our own